File: clients/scala/src/main/scala/util/StreamUtil.scala
package util

import java.io.{IOException, InputStream, OutputStream}
import java.nio.charset.StandardCharsets
import java.nio.{ByteBuffer, ByteOrder}


object StreamUtil {

  def readBytes(stream: InputStream, byteCount: Int): Array[Byte] = {
    val bytes = new Array[Byte](byteCount)
    var offset = 0
    while (offset < bytes.size) {
      val read = stream.read(bytes, offset, bytes.size - offset)
      if (read == -1) {
        return bytes
      }
      offset += read
    }
    if (offset != bytes.size) {
      throw new IOException("Unexpected EOF")
    }
    bytes
  }

  def readBoolean(stream: InputStream): Boolean = {
    ByteBuffer.wrap(readBytes(stream, 1)).get() != 0
  }

  def readInt(stream: InputStream): Int = {
    ByteBuffer.wrap(readBytes(stream, Integer.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getInt
  }

  def readLong(stream: InputStream): Long = {
    ByteBuffer.wrap(readBytes(stream, java.lang.Long.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getLong
  }

  def readFloat(stream: InputStream): Float = {
    ByteBuffer.wrap(readBytes(stream, java.lang.Float.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getFloat
  }

  def readDouble(stream: InputStream): Double = {
    ByteBuffer.wrap(readBytes(stream, java.lang.Double.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getDouble
  }

  def readString(stream: InputStream): String = {
    new String(readBytes(stream, readInt(stream)), StandardCharsets.UTF_8)
  }

  def writeBytes(stream: OutputStream, bytes: Array[Byte]): Unit = {
    stream.write(bytes)
  }

  def writeBoolean(stream: OutputStream, value: Boolean): Unit = {
    writeBytes(stream, Array.fill[Byte](1)((if (value) 1 else 0).toByte))
  }

  def writeInt(stream: OutputStream, value: Int): Unit = {
    writeBytes(stream, ByteBuffer.allocate(Integer.BYTES).order(ByteOrder.LITTLE_ENDIAN).putInt(value).array())
  }

  def writeLong(stream: OutputStream, value: Long): Unit = {
    writeBytes(stream, ByteBuffer.allocate(java.lang.Long.BYTES).order(ByteOrder.LITTLE_ENDIAN).putLong(value).array())
  }

  def writeFloat(stream: OutputStream, value: Float): Unit = {
    writeBytes(stream, ByteBuffer.allocate(java.lang.Float.BYTES).order(ByteOrder.LITTLE_ENDIAN).putFloat(value).array())
  }

  def writeDouble(stream: OutputStream, value: Double): Unit = {
    writeBytes(stream, ByteBuffer.allocate(java.lang.Double.BYTES).order(ByteOrder.LITTLE_ENDIAN).putDouble(value).array())
  }

  def writeString(stream: OutputStream, value: String): Unit = {
    val bytes = value.getBytes(StandardCharsets.UTF_8)
    writeInt(stream, bytes.size)
    writeBytes(stream, bytes)
  }
}

File: clients/java/src/main/java/util/StreamUtil.java
package util;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

public class StreamUtil {
    public static byte[] readBytes(InputStream stream, int byteCount) throws IOException {
        byte[] bytes = new byte[byteCount];
        int offset = 0;
        while (offset < bytes.length) {
            int read = stream.read(bytes, offset, bytes.length - offset);
            if (read == -1) {
                break;
            }
            offset += read;
        }
        if (offset != bytes.length) {
            throw new IOException("Unexpected EOF");
        }
        return bytes;
    }

    public static boolean readBoolean(InputStream stream) throws IOException {
        return ByteBuffer.wrap(readBytes(stream, 1)).get() != 0;
    }

    public static int readInt(InputStream stream) throws IOException {
        return ByteBuffer.wrap(readBytes(stream, Integer.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getInt();
    }

    public static long readLong(InputStream stream) throws IOException {
        return ByteBuffer.wrap(readBytes(stream, Long.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getLong();
    }

    public static float readFloat(InputStream stream) throws IOException {
        return ByteBuffer.wrap(readBytes(stream, Float.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getFloat();
    }

    public static double readDouble(InputStream stream) throws IOException {
        return ByteBuffer.wrap(readBytes(stream, Double.BYTES)).order(ByteOrder.LITTLE_ENDIAN).getDouble();
    }

    public static String readString(InputStream stream) throws IOException {
        int length = readInt(stream);
        return new String(readBytes(stream, length), StandardCharsets.UTF_8);
    }

    public static void writeBytes(OutputStream stream, byte[] bytes) throws IOException {
        stream.write(bytes);
    }

    public static void writeBoolean(OutputStream stream, boolean value) throws IOException {
        writeBytes(stream, new byte[] { (byte) (value ? 1 : 0) });
    }

    public static void writeInt(OutputStream stream, int value) throws IOException {
        writeBytes(stream, ByteBuffer.allocate(Integer.BYTES).order(ByteOrder.LITTLE_ENDIAN).putInt(value).array());
    }

    public static void writeLong(OutputStream stream, long value) throws IOException {
        writeBytes(stream, ByteBuffer.allocate(Long.BYTES).order(ByteOrder.LITTLE_ENDIAN).putLong(value).array());
    }

    public static void writeFloat(OutputStream stream, float value) throws IOException {
        writeBytes(stream, ByteBuffer.allocate(Float.BYTES).order(ByteOrder.LITTLE_ENDIAN).putFloat(value).array());
    }

    public static void writeDouble(OutputStream stream, double value) throws IOException {
        writeBytes(stream, ByteBuffer.allocate(Double.BYTES).order(ByteOrder.LITTLE_ENDIAN).putDouble(value).array());
    }

    public static void writeString(OutputStream stream, String value) throws IOException {
        byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
        writeInt(stream, bytes.length);
        writeBytes(stream, bytes);
    }
}